dev blog, needs log and maybe some editing. (#2025)
FusionAuth/fusionauth-site: site/_posts/2023-03-10-growing-is-hard.md

---
layout: blog-post

> title: Building a team is hard&#58; part I

Reviewer: Is part 2 about the technology and domain? because those are worth addressing (our particular tech stack, java in general is not as sexy, the auth domain has unique challenges...)

Author: not sure what part 2 is yet.. but figured it would be something. :-)

description: Building a team is harder than it looks. The attrition problem.
author: Daniel DeGroff
category: article

> image: blogs/consents-app/managing-consent-header.png

Reviewer 2: Assuming "we'll need a graphic" means this is a placeholder.

Author: Yes. Sean will help me with a better graphic, this is just a placeholder.

tags: startup bootstrap dev hiring firing attrition
excerpt_separator: "<!--more-->"
---

> Building a engineering team is hard. There are many reasons for this, and one of them is attrition. In this article I'll talk about my pain and suffering relate to attrition while building FusionAuth. 

Reviewer 2: ```suggestion
Building an engineering team is hard. There are many reasons for this, and one of them is attrition. In this article I'll talk about my pain and suffering related to attrition while building FusionAuth. 
```

<!--more-->

### Statistics

> The average tenure of a software engineer varies. Interestingly it may even be the lowest in the largest tech companies such as Facebook (Meta), Amazon, Apple, Netflix and Google, commonly referred to as "FAANG".

Reviewer 2: ```suggestion
The average tenure of a software engineer varies. Interestingly, tenure may be lowest in the largest tech companies such as Facebook (Meta), Amazon, Apple, Netflix, and Google, commonly referred to as "FAANG".
```

Using "it" confused me a bit. The article is about attrition. The antecedent is unclear.

Is "FAANG" still common parlance? I remember this from 10 years ago, but I haven't heard it recently. Granted, I don't follow tech media too closely.

Reviewer 2: > Use the oxford comma. Apples, bananas, and oranges are my favorite fruits.

Author: We could delete the `FAANG` reference - it is still used but maybe antiquated. And it would technically "MAANG" now. :-)

> One report shows an average tenure in big tech is between 1.1 and 3.2 years.[^1] The US Bureau of Labor and Statistics report a nationwide average of 3.8 to 4.3 years which is not limited to the software industry.[^2] Regardless of which statistic you use, it is hard to argue that average tenure is short, and almost certainly lower than the average for the general US population.

Reviewer 2: ```suggestion
One report shows an average tenure in big tech is between 1.1 and 3.2 years.[^1] The US Bureau of Labor and Statistics report a nationwide average of 3.8 to 4.3 years which is not limited to the software industry.[^2] Regardless of which statistic you use, it is hard to argue that average tenure in big tech is short, and almost certainly lower than the average across all sectors.
```

Not sure on the wording at the end, but the average is not across individuals. It's across different areas of the economy.

### Bye, Felicia
Switching jobs means new experiences, more money and a change of scenery. This can be very appealing early in your career and can help you level up quickly.

> The downside is that you rarely develop the same level of experiences and skill that you would by sticking around, learning to work through conflict, mastering a code base and learning how to stay productive in spite of company politics.

Reviewer 2: ```suggestion
The downside is that you rarely develop the same level of experiences and skill that you would by sticking around, learning to work through conflict, mastering a code base, and learning how to stay productive in spite of company politics.
```

> Use the oxford comma. Apples, bananas, and oranges are my favorite fruits.

> The downside is that you rarely develop the same level of experiences and skill that you would by sticking around, learning to work through conflict, mastering a code base, and learning how to stay productive in spite of company politics.

Reviewer: ```suggestion
The downside of switching jobs regularly is that you rarely develop the experience and skill that you would by sticking around, learning to work through conflict, mastering a code base, and learning how to stay productive in spite of company politics. There's a rigor in living in a project for years, as you see it evolve over time and can learn when architectural choices add flexibility and when they constrain. Sure, you can learn that from books, but seeing it in real life drives home the importance. Additionally, when working in an existing code base, you gain humility. There's nothing quite so humbling as fixing a bug you introduced a year or two ago: "what on earth was I thinking!".
```

Unfortunately this doesn't seem to be the sentiment shared by many engineers. 

### The problem
By the time an employee is beginning to produce and become competent - they have one foot out the door looking for their next adventure. 

> There are various ways to overcome this problem.

Reviewer: ```suggestion
This is an issue, because you want employees to be productive for years, not months. There are various approaches to solve this problem.
```

> Venture capital backed companies solve this by hiring many engineers at once. Using this strategy, even if a few leave, there is nearly always overlap in tenure so that the combined skills on the team never go to zero.

Reviewer 2: ```suggestion
Venture Capital-backed companies solve this by hiring many engineers at once. Using this strategy, even if a few leave, there is nearly always overlap in tenure so that the combined skills on the team never go to zero.
```

> Venture capital backed companies solve this by hiring many engineers at once. Using this strategy, even if a few leave, there is nearly always overlap in tenure so that the combined skills on the team never go to zero.

Reviewer: ```suggestion
Venture capital backed companies solve this by hiring many engineers at once. They all learn the code base and can support each other during the initial time. If there are existing engineers, they can teach all their team mates at once. Using this strategy, even if a few folks leave early, there overlap in tenure. Team members can pass skills on through the cultural osmosis that is so important at augmenting documentation.
```

> At FusionAuth, we've taken a different path. things a little different. We've never taken capital, we've built a profitable business w/out debt. We hire when we can afford to do so.

Reviewer 2: ```suggestion
At FusionAuth, we've taken a different path. Things are a little different. We've never taken capital, we've built a profitable business without debt, and we hire when we can afford to do so.
```

> At FusionAuth, we've taken a different path. things a little different. We've never taken capital, we've built a profitable business w/out debt. We hire when we can afford to do so.

Reviewer: ```suggestion
At FusionAuth, we've taken a different path. We've never taken capital, we've built a profitable business without debt, and we hire when we can afford to do so. This means that we don't hire cohorts of engineers. 
```

> Hiring slow, means we may only be able to afford to hire a new engineer every 12-18 months. If the average tenure is less than your hiring cadence - you'll never get ahead of the attrition curve.

Reviewer 2: ```suggestion
Hiring slow means we may only be able to afford to hire a new engineer every 12-18 months. If the average tenure is less than your hiring cadence, you'll never get ahead of the attrition curve.
```

> Hiring slow, means we may only be able to afford to hire a new engineer every 12-18 months. If the average tenure is less than your hiring cadence - you'll never get ahead of the attrition curve.

Reviewer: ```suggestion
Bootstrapping means we may only be able to afford to add an additional engineering position. If the average developer tenure is less than the hiring timeframe, you never get ahead of the attrition curve. An engineer joins, goes from not knowing much about the product and code base, and then leaves. The next engineer follows the same pattern. And the next and the next.
```

> It is sort of like filling a tub without plugging the drain. Unless you have great water pressure , the tub will drain faster than you can fill it. It is just difficult to get ahead, and you end up sitting cold and naked in the tub. The analogy does fall apart a bit at the end.   

Reviewer 2: ```suggestion
It's sort of like filling a bathtub without plugging the drain. Unless you have great water pressure, the bathtub will drain faster than you can fill it. It is difficult to get ahead, and you end up sitting cold and naked in the bathtub. The analogy does fall apart a bit at the end.
```

The tone goes back and forth. I think "tub" is too broad a term for a blog post and should be more specific.

> It is sort of like filling a tub without plugging the drain. Unless you have great water pressure , the tub will drain faster than you can fill it. It is just difficult to get ahead, and you end up sitting cold and naked in the tub. The analogy does fall apart a bit at the end.   

Reviewer: ```suggestion
It is sort of like filling a bathtub without plugging the drain. Unless you have exceptional water pressure, the tub will drain faster than you can fill it. It is impossible to get ahead, and if you get in, you end up sitting cold and naked in the tub. (The analogy does fall apart a bit at the end, but both are unpleasant experiences.)
```

> The result is that it leaves the technical founders holding the bag every time. This cycle repeats and can become demoralizing when the time and effort that is spent imparting the skills necessary to continue and grow the business are never preserved.

Reviewer: ```suggestion
The result of the inability to retain engineers is that the technical founders hold the bag every time a technical issue arises. If there's a hard problem, they are the fallback. If there's a technical sales call, they are pulled in. If there's a support issue in the middle of the night for a customer, they are woken up.

It gets worse!

This demoralizing cycle repeats. The time and effort that is spent teaching the skills necessary to continue improving the code base and therefore grow the business are expended again and again with every new engineer. Each time the technical founders hope that this engineer is the one that will stick around. This engineer is the one that they can teach and partner with to improve the product. This engineer will become part of the core team.
```

> 

Reviewer: ```suggestion

### Remedies

Talk about possible remedies that don't involve hiring more engineers and why they didn't work.
documetnation
automation
pay more 
make it a nicer place to work so people didn't leave
```

### Light at the end of the tunnel

> Thankfully, we have turned the corner on this issue and begun to dig our way out of the hole. As FusionAuth has continued to grow, we can now afford to pay competitive salaries, and we also have enough revenue to hire in bulk. Hiring more than one engineer at a time is helpful in several ways.

Reviewer: ```suggestion
Thankfully, we have turned the corner on this issue and begun to dig our way out of the hole. As FusionAuth has continued to grow, we can now afford to pay competitive salaries, and we also have enough revenue to hire multiple engineers at once. Hiring more than one engineer at a time is helpful in several ways.
```

> Thankfully, we have turned the corner on this issue and begun to dig our way out of the whole. As FusionAuth can continued to grow, we can now afford to pay competitive salaries, and we also have enough revenue to hire in bulk. Hiring more than one engineer at a time is helpful in several ways. 

Reviewer 2: ```suggestion
Thankfully, we have turned the corner on this issue and begun to dig our way out of the hole. As FusionAuth has continued to grow, we can now afford to pay competitive salaries, and we also have enough revenue to hire in bulk. Hiring more than one engineer at a time is helpful in several ways.
```

> The first is that the new team feels a sense of community as they all start from zero together. They do not have to compare their skill set or knowledge of the product with someone that has been doing it for 10 years, and instead can compare to another that is learning the same things.

Reviewer 2: ```suggestion
The first is that the new team feels a sense of camaraderie as they all start from "zero" together. They do not have to compare their skillset or knowledge of the product with someone that has been working in the space for 10 years, and instead can compare to another that is starting from the same place at the same time.
```

The "compare to another" also sounds like it's a competition. Not sure if there is a better way to phrase that.

> The first is that the new team feels a sense of camaraderie as they all start from "zero" together. They do not have to compare their skillset or knowledge of the product with someone that has been working in the space for 10 years, and instead can compare to another that is starting from the same place at the same time.

Reviewer: ```suggestion
The first is that the new team feels a sense of camaraderie as they all start exploring and working in the new code base together. They don't have any experience in the product, but that's okay. They do not have to compare their skillset or knowledge of the product with someone that has been working in the space for 10 years and feel despondent. Instead they benchmark against one another, and of course everyone in this cohort is starting with similar knowledge.
```

> The second - which is most exciting to me - is that we are now filling the tub four times as fast!

Reviewer 2: ```suggestion
The second - which is most exciting to me - is that we are now filling the bathtub four times as fast!
```

Also, if we're going with math, it's only 3 times as fast.

Author: I sort of wanted to say tub to make it more slang.

### Lessons learned

Seeing mistakes in hindsight is not difficult. I would do many things differently given the opportunity. 

> One of our biggest mistakes is thinking we could not afford to hire senior engineers. It can be difficult to hire these engineers early in a company because their salary requirements will likely be more than you pay yourself. But the reality is that they can produce 5-10x of an entry level engineer, and their salary not even close to 5-10x, and more likely 2x. 

Reviewer 2: I'm not a fan of the productivity comparison. Partly because I think the "10x engineer" thing is dumb.

Author: This isn't so much a "10x-er" statement - but the reality of what an experienced engineer can do in contrast to a brand new engineer. Open to better ways to say it -but there certainly is a difference in productivity.

Reviewer: I'd anchor this in personal experience, and maybe leave off the numbers (which I assume are anecdotal). Maybe something like:

```
But the reality is that a senior engineer can get up to speed more quickly and therefore can produce far more value (through code and processes) than someone with less experience. The value they bring is far more than the extra cost. For example, we had a senior engineer take on a task similar to <X>, whereas before a less senior engineer had a hard time doing a similar task.
```

Reviewer: Not sure about if you are going to call out anyone, so maybe ignore the 'for example' above if you can't do so without saying unkind things about someone.

> That is to say - the cost benefit analysis clearly indicates that paying more for experience is worth it. That isn't to say you shouldn't hire people of all experience levels, but when you are building a company from scratch- you want to be hiring people that you want to be leading the company in 2-3 years.

Reviewer 2: ```suggestion
The cost benefit analysis clearly indicates that paying more for experience is worth it. That isn't to say you shouldn't hire people of all experience levels, but when you are building a company from scratch, you want to be hiring people that you want to be leading the company in 2-3 years.
```

-----

> [^1]: https://developerpitstop.com/how-long-do-software-engineers-stay-at-a-job/

Reviewer: I'd inline these.
